## iPinYou DataSet EDA 

In [1]:
# imports 
import pandas as pd 
import numpy as np

In [2]:
# loading the data 
from ad_causal_doubleml.config.paths import DATA_DIR
file_path = DATA_DIR / "../1458/train.log.txt"

# explicit dtypes reduce memory usage when loading
# the large dataset
dtypes = {
    "click": "int8",
    "weekday": "int8",
    "hour": "int8",
    "logtype": "int8",

    "region": "int16",
    "city": "int16",

    "adexchange": "category",

    "slotwidth": "int16",
    "slotheight": "int16",

    "slotvisibility": "category",
    "slotformat": "int16",

    "slotprice": "int16",

    "bidprice": "int16",
    "payprice": "int16",

    "advertiser": "category",

    # High-cardinality categorical features
    "bidid": "string",
    "timestamp": "string",
    "ipinyouid": "string",
    "useragent": "category",
    "IP": "string",
    "domain": "string",
    "url": "string",
    "urlid": "string",
    "slotid": "string",
    "creative": "string",
    "keypage": "string",
    "usertag": "string"
}

df = pd.read_csv(
    file_path,
    sep="\t",
    dtype=dtypes,
    na_values=["null"]
)

#### Data Shape and Dtypes

In [3]:
print(df.shape)
print(df.dtypes)

(3083056, 27)
click                       int8
weekday                     int8
hour                        int8
bidid             string[python]
timestamp         string[python]
logtype                     int8
ipinyouid         string[python]
useragent               category
IP                string[python]
region                     int16
city                       int16
adexchange              category
domain            string[python]
url               string[python]
urlid             string[python]
slotid            string[python]
slotwidth                  int16
slotheight                 int16
slotvisibility          category
slotformat                 int16
slotprice                  int16
creative          string[python]
bidprice                   int16
payprice                   int16
keypage           string[python]
advertiser              category
usertag           string[python]
dtype: object


#### Missing Values 

In [4]:
df.isna().sum().sort_values(ascending=False)

urlid             3083056
usertag            399463
domain             169140
url                 68566
timestamp               0
slotwidth               0
advertiser              0
keypage                 0
payprice                0
bidprice                0
creative                0
slotprice               0
slotformat              0
slotvisibility          0
slotheight              0
slotid                  0
logtype                 0
hour                    0
weekday                 0
bidid                   0
adexchange              0
city                    0
region                  0
IP                      0
useragent               0
ipinyouid               0
click                   0
dtype: int64

Percentage of a feature that is missing: 

In [5]:
(df.isna().mean() * 100).sort_values(ascending=False).round(2)

urlid             100.00
usertag            12.96
domain              5.49
url                 2.22
timestamp           0.00
slotwidth           0.00
advertiser          0.00
keypage             0.00
payprice            0.00
bidprice            0.00
creative            0.00
slotprice           0.00
slotformat          0.00
slotvisibility      0.00
slotheight          0.00
slotid              0.00
logtype             0.00
hour                0.00
weekday             0.00
bidid               0.00
adexchange          0.00
city                0.00
region              0.00
IP                  0.00
useragent           0.00
ipinyouid           0.00
click               0.00
dtype: float64

#### Duplicate Records 

In [6]:
# the number of completely identical records
print(df.duplicated().sum())
df[df.duplicated()]

1


,click,weekday,hour,bidid,timestamp,logtype,ipinyouid,useragent,IP,region,...,slotheight,slotvisibility,slotformat,slotprice,creative,bidprice,payprice,keypage,advertiser,usertag
1026676,0,6,10,ac2a4c639cbac54d3a2aca8576441621,20130608105218564,1,VhkxO5xALtk4FWc,windows_ie,218.93.21.*,80,...,90,2,0,5,48f2e9ba15708c0146bda5e1dd653caa,300,60,bebefa5efe83beee17a3d245e7c5085b,1458,<NA>


In [7]:
# the duplicated entry 
df[df["bidid"]=="ac2a4c639cbac54d3a2aca8576441621"]
# can see that usertage entry is "<NA>"

,click,weekday,hour,bidid,timestamp,logtype,ipinyouid,useragent,IP,region,...,slotheight,slotvisibility,slotformat,slotprice,creative,bidprice,payprice,keypage,advertiser,usertag
1024054,0,6,10,ac2a4c639cbac54d3a2aca8576441621,20130608104910589,1,VhkxO5xALtk4FWc,windows_ie,218.93.21.*,80,...,90,2,0,5,48f2e9ba15708c0146bda5e1dd653caa,300,60,bebefa5efe83beee17a3d245e7c5085b,1458,<NA>
1025722,0,6,10,ac2a4c639cbac54d3a2aca8576441621,20130608105007505,1,VhkxO5xALtk4FWc,windows_ie,218.93.21.*,80,...,90,2,0,5,48f2e9ba15708c0146bda5e1dd653caa,300,60,bebefa5efe83beee17a3d245e7c5085b,1458,<NA>
1025723,0,6,10,ac2a4c639cbac54d3a2aca8576441621,20130608105048784,1,VhkxO5xALtk4FWc,windows_ie,218.93.21.*,80,...,90,2,0,5,48f2e9ba15708c0146bda5e1dd653caa,300,60,bebefa5efe83beee17a3d245e7c5085b,1458,<NA>
1025724,0,6,10,ac2a4c639cbac54d3a2aca8576441621,20130608105051022,1,VhkxO5xALtk4FWc,windows_ie,218.93.21.*,80,...,90,2,0,5,48f2e9ba15708c0146bda5e1dd653caa,300,60,bebefa5efe83beee17a3d245e7c5085b,1458,<NA>
1026513,0,6,10,ac2a4c639cbac54d3a2aca8576441621,20130608105133520,1,VhkxO5xALtk4FWc,windows_ie,218.93.21.*,80,...,90,2,0,5,48f2e9ba15708c0146bda5e1dd653caa,300,60,bebefa5efe83beee17a3d245e7c5085b,1458,<NA>
1026514,0,6,10,ac2a4c639cbac54d3a2aca8576441621,20130608105133532,1,VhkxO5xALtk4FWc,windows_ie,218.93.21.*,80,...,90,2,0,5,48f2e9ba15708c0146bda5e1dd653caa,300,60,bebefa5efe83beee17a3d245e7c5085b,1458,<NA>
1026522,0,6,10,ac2a4c639cbac54d3a2aca8576441621,20130608105147074,1,VhkxO5xALtk4FWc,windows_ie,218.93.21.*,80,...,90,2,0,5,48f2e9ba15708c0146bda5e1dd653caa,300,60,bebefa5efe83beee17a3d245e7c5085b,1458,<NA>
1026523,0,6,10,ac2a4c639cbac54d3a2aca8576441621,20130608105147104,1,VhkxO5xALtk4FWc,windows_ie,218.93.21.*,80,...,90,2,0,5,48f2e9ba15708c0146bda5e1dd653caa,300,60,bebefa5efe83beee17a3d245e7c5085b,1458,<NA>
1026525,0,6,10,ac2a4c639cbac54d3a2aca8576441621,20130608105159242,1,VhkxO5xALtk4FWc,windows_ie,218.93.21.*,80,...,90,2,0,5,48f2e9ba15708c0146bda5e1dd653caa,300,60,bebefa5efe83beee17a3d245e7c5085b,1458,<NA>
1026526,0,6,10,ac2a4c639cbac54d3a2aca8576441621,20130608105159243,1,VhkxO5xALtk4FWc,windows_ie,218.93.21.*,80,...,90,2,0,5,48f2e9ba15708c0146bda5e1dd653caa,300,60,bebefa5efe83beee17a3d245e7c5085b,1458,<NA>


Remove the row that is completely duplicated: 

In [8]:
df = df[df["bidid"] != "ac2a4c639cbac54d3a2aca8576441621"]

The number of remaining bid ids

In [9]:
# dataframe containing only duplicated entries
df_duplicated = df[df["bidid"].duplicated(keep=False)]


In [10]:
df_duplicated.shape

(11619, 27)

In [11]:
df_duplicated["bidid"].unique().shape
# this implies duplicate bidids have many rows
# this is the number of rows I aim to end up with
# one row for each unique bidid

(5602,)

Check for empty strings

In [12]:
# check for empty strings 
"""
obj_cols = df.select_dtypes(include='object').columns

(df
    .apply(lambda s: s.astype(str).str.strip().eq('').sum())
    .sort_values(ascending=False))
    """

"\nobj_cols = df.select_dtypes(include='object').columns\n\n(df\n    .apply(lambda s: s.astype(str).str.strip().eq('').sum())\n    .sort_values(ascending=False))\n    "

In [13]:
compare_cols = [c for c in df_duplicated.columns if c != "bidid"]
results = []

# only process duplicated bidid groups 


In [14]:
dup_groups = df_duplicated.groupby("bidid").filter(lambda g: len(g) > 1).groupby("bidid")

In [15]:
for bidid, group in dup_groups:
    # Columns where values differ within this bidid
    diff_cols = [col for col in compare_cols if group[col].nunique(dropna=False) > 1]

    if diff_cols:
        # Store differing columns and their values
        diff_values = {
            col: group[col].tolist()
            for col in diff_cols
        }

        results.append({
            "bidid": bidid,
            "different_features": diff_cols,
            "values": diff_values
        })

# Convert to dataframe for easy viewing
diff_df = pd.DataFrame(results)

In [16]:
#print(diff_df)

dict = {}

total_diff_in_timestamps = 0
num_diff_timestamps = 0

from datetime import datetime
# Parse: first 14 digits are yyyyMMddHHmmss, last 3 are milliseconds
def parse_ts(ts: str) -> datetime:
    return datetime.strptime(ts, '%Y%m%d%H%M%S%f')


# Optional: pretty print each bidid
for _, row in diff_df.iterrows():
    print(f"\nBidID: {row['bidid']}")
    print("Different features:", row["different_features"])
    for feature in row["different_features"]:
        if feature in dict:
            dict[feature] += 1
        else:
            dict[feature] = 1
    for feature, values in row["values"].items():
        print(f"  {feature}: {values}")
        if feature == "timestamp":
            #print(int(values[1])-int(values[0]))

            dt1 = parse_ts(values[0])
            dt2 = parse_ts(values[1]) 
            if len(values) > 2:
                print("number of timestamp values more than two")
                break
            delta_s = (dt2 - dt1).total_seconds()

            total_diff_in_timestamps += delta_s
            num_diff_timestamps += 1



BidID: 10017f7b042216349d89dfe7d9ada3a7
Different features: ['timestamp', 'ipinyouid']
  timestamp: ['20130612141014886', '20130612141014982']
  ipinyouid: ['Vh5zPnqmLQb9FBa', 'Vh5zP3p63U8fjQc']

BidID: 1007fdd9a74414192a03c9bd77202a0a
Different features: ['timestamp', 'ipinyouid', 'usertag']
  timestamp: ['20130612204207951', '20130612204208020']
  ipinyouid: ['Vh5zLxl83Q1VJVR', 'Vh5zLxl9OHEIjZa']
  usertag: [<NA>, '10063,10111']

BidID: 100df7877c81acfdf999e1a0ee1eb73f
Different features: ['timestamp', 'ipinyouid', 'usertag']
  timestamp: ['20130612152514254', '20130612152514288']
  ipinyouid: ['Vhk0CDpqD4xUFGn', 'Vh5zPaK6P2F6qhc']
  usertag: ['10006,10110', '10102,10063,10006,10110']

BidID: 1014d5f55911b7186f9af6462c9911f0
Different features: ['timestamp', 'ipinyouid', 'usertag']
  timestamp: ['20130610000104347', '20130610000104486']
  ipinyouid: ['VhkyOaCx3QTLjMR', 'Vh5hZAnxP9T2Xh2']
  usertag: ['10063', <NA>]

BidID: 1014ef29721f3210bbd540fd52e50a3f
Different features: ['timest

In [17]:
dict

{'timestamp': 5584,
 'ipinyouid': 4146,
 'usertag': 2590,
 'useragent': 180,
 'hour': 48,
 'weekday': 1,
 'payprice': 8}

Removing null usertags

In [18]:
# seperating the rows into where usertag is null 
df_duplicated_temp = df_duplicated
print(df_duplicated_temp.shape)

df_usertag_null = df_duplicated_temp[df_duplicated_temp["usertag"].isna()]
print(f"The number of null usertags is {df_usertag_null['bidid'].count()}")
df_usertag_not_null = df_duplicated_temp[df_duplicated_temp["usertag"].notna()]
# retain nulls where bid id is in not nulls
df_usertag_null = df_usertag_null[df_usertag_null['bidid'].isin(df_usertag_not_null['bidid'])]
print(f"The number of null usertags where there is a non-null usertag is: {df_usertag_null.shape[0]}")


# drop all rows in df_duplicated_temp that appear in df_usertag_null
result = df_duplicated_temp.merge(df_usertag_null.drop_duplicates(), how='left', indicator=True)
df_duplicated_temp = result = result[result['_merge'] == 'left_only'].drop(columns='_merge')
df_duplicated_temp.shape

(11619, 27)
The number of null usertags is 4458
The number of null usertags where there is a non-null usertag is: 1841


(9778, 27)

In [19]:
df_duplicated = df_duplicated_temp

Where the only difference between rows is the timestamp and ipinyouid, retain the row with the later timestamp

In [20]:
import pandas as pd

# Make a copy
df_clean = df_duplicated.copy()

# Convert timestamp to datetime
df_clean["timestamp_dt"] = pd.to_datetime(
    df_clean["timestamp"].astype(str),
    format="%Y%m%d%H%M%S%f"
)

# Columns that must be identical for two rows to be considered duplicates
compare_cols = [
    col for col in df_clean.columns
    if col not in ["timestamp", "ipinyouid", "timestamp_dt"]
]

# Sort by timestamp so the newest row comes first
df_clean = df_clean.sort_values("timestamp_dt", ascending=False)

# Remove duplicates where every column except timestamp and iPinYouid
# is identical. The newest timestamp is retained.
df_clean = df_clean.drop_duplicates(
    subset=compare_cols,
    keep="first"
)

# Remove helper column
df_clean = df_clean.drop(columns="timestamp_dt")

# Reset index
df_clean = df_clean.reset_index(drop=True)

In [21]:
print(f"The number of rows removed where the only difference was the timestamp \nand ipinyouid: {df_duplicated.shape[0]-df_clean.shape[0]}. The older timestamp record is retained.")

The number of rows removed where the only difference was the timestamp 
and ipinyouid: 3354. The older timestamp record is retained.


In [22]:
df_duplicated = df_clean
df_duplicated.shape

(6424, 27)

Average difference in timestamps when there are more than one timestamp:

In [23]:
print(f"Average difference in timestamps when there is more than one \ntimestemp for a given bid id: {total_diff_in_timestamps / num_diff_timestamps}.")

Average difference in timestamps when there is more than one 
timestemp for a given bid id: 20.289536302395234.


In [24]:
# Parse function
def parse_ts(ts):
    return datetime.strptime(ts, '%Y%m%d%H%M%S%f')

# Convert all timestamps
dts = [parse_ts(ts) for ts in df["timestamp"]]

# Find min and max
min_dt = min(dts)
max_dt = max(dts)

# Difference in seconds
diff_seconds = (max_dt - min_dt).total_seconds()

print('Min:', min_dt)
print('Max:', max_dt)
print('Difference (seconds):', diff_seconds)

Min: 2013-06-06 00:01:04.828000
Max: 2013-06-12 23:42:04.173000
Difference (seconds): 603659.345


In [25]:
results = []
dup_groups = df_duplicated.groupby("bidid").filter(lambda g: len(g) > 1).groupby("bidid")
for bidid, group in dup_groups:
    # Columns where values differ within this bidid
    diff_cols = [col for col in compare_cols if group[col].nunique(dropna=False) > 1]

    if diff_cols:
        # Store differing columns and their values
        diff_values = {
            col: group[col].tolist()
            for col in diff_cols
        }

        results.append({
            "bidid": bidid,
            "different_features": diff_cols,
            "values": diff_values
        })

# Convert to dataframe for easy viewing
diff_df = pd.DataFrame(results)

#print(diff_df)

dict = {}

total_diff_in_timestamps = 0
num_diff_timestamps = 0

from datetime import datetime
# Parse: first 14 digits are yyyyMMddHHmmss, last 3 are milliseconds
def parse_ts(ts: str) -> datetime:
    return datetime.strptime(ts, '%Y%m%d%H%M%S%f')


# Optional: pretty print each bidid
for _, row in diff_df.iterrows():
    print(f"\nBidID: {row['bidid']}")
    print("Different features:", row["different_features"])
    for feature in row["different_features"]:
        if feature in dict:
            dict[feature] += 1
        else:
            dict[feature] = 1
    for feature, values in row["values"].items():
        print(f"  {feature}: {values}")



BidID: 100df7877c81acfdf999e1a0ee1eb73f
Different features: ['usertag']
  usertag: ['10102,10063,10006,10110', '10006,10110']

BidID: 10a979105df92f827a8c1cbffa4efa39
Different features: ['usertag']
  usertag: ['10063', '10063,10110']

BidID: 10b4506325bac95dfbf6269619288280
Different features: ['usertag']
  usertag: ['10006,10110', '10059,10063,10006,10110']

BidID: 110625affedf5dc3d003aec8c7d0a616
Different features: ['usertag']
  usertag: ['10063,10006,10110', '10006']

BidID: 1112543bd4ed952865835944d9f6ed32
Different features: ['usertag']
  usertag: ['10063,13403,10110', '10063,10006']

BidID: 112ee8cd03df99bd97f1d163e2445200
Different features: ['useragent']
  useragent: ['other_other', 'windows_ie']

BidID: 116821a3b71a4831fc0e24aa5caaea13
Different features: ['usertag']
  usertag: ['10006,10110', '10006,10059,13403,10063,10110']

BidID: 11856f50072a93eec2098ca072a8d294
Different features: ['usertag']
  usertag: ['10083,10057,10006,14273,10110', '10006,10052,10057,10063,10083,1

In [26]:
dict

{'usertag': 756, 'useragent': 134, 'hour': 37, 'payprice': 6}

In [27]:
import numpy as np

# Randomly select one row for each bidid
random_rows = df_duplicated.groupby("bidid").sample(n=1, random_state=42)

# Combine all usertags for each bidid
combined_tags = (
    df_duplicated.groupby("bidid")["usertag"]
      .apply(lambda x: ",".join(sorted(set(
          tag
          for value in x.dropna()
          for tag in value.split(",")
      ))))
)

# Replace the usertag in the randomly selected rows
random_rows["usertag"] = random_rows["bidid"].map(combined_tags)

# Result
df_combined = random_rows.reset_index(drop=True)

In [28]:
df_combined # shouldn't we be left with 5,000+, the number of unique bids
# out of the duplicates

,click,weekday,hour,bidid,timestamp,logtype,ipinyouid,useragent,IP,region,...,slotheight,slotvisibility,slotformat,slotprice,creative,bidprice,payprice,keypage,advertiser,usertag
0,0,3,14,10017f7b042216349d89dfe7d9ada3a7,20130612141014982,1,Vh5zP3p63U8fjQc,windows_chrome,106.86.23.*,275,...,280,0,1,0,77819d3e0b3467fe5c7b16d68ad923a1,300,151,bebefa5efe83beee17a3d245e7c5085b,1458,
1,0,3,20,1007fdd9a74414192a03c9bd77202a0a,20130612204208020,1,Vh5zLxl9OHEIjZa,windows_chrome,218.75.146.*,201,...,90,1,0,5,48f2e9ba15708c0146bda5e1dd653caa,300,15,bebefa5efe83beee17a3d245e7c5085b,1458,"10063,10111"
2,0,3,15,100df7877c81acfdf999e1a0ee1eb73f,20130612152514254,1,Vhk0CDpqD4xUFGn,windows_chrome,125.85.66.*,275,...,280,2,0,4,77819d3e0b3467fe5c7b16d68ad923a1,300,4,bebefa5efe83beee17a3d245e7c5085b,1458,"10006,10063,10102,10110"
3,0,1,0,1014d5f55911b7186f9af6462c9911f0,20130610000104347,1,VhkyOaCx3QTLjMR,windows_chrome,101.85.56.*,79,...,280,0,1,0,77819d3e0b3467fe5c7b16d68ad923a1,300,77,bebefa5efe83beee17a3d245e7c5085b,1458,10063
4,0,0,20,1014ef29721f3210bbd540fd52e50a3f,20130609200033260,1,VhkyLuMoD812Qhk,windows_ie,221.214.216.*,146,...,250,1,0,5,2abc9eaf57d17a96195af3f63c45dc72,300,77,bebefa5efe83beee17a3d245e7c5085b,1458,13866
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5597,0,0,14,ffb7f078a336d4b4868973c158dcf9a0,20130609142515038,1,VhkROu55ONFc3oL,windows_ie,110.240.218.*,3,...,280,2,0,5,77819d3e0b3467fe5c7b16d68ad923a1,300,157,bebefa5efe83beee17a3d245e7c5085b,1458,"10024,10076"
5598,0,2,2,ffc9d07883e6c681c69e87bb85b1622b,20130611022107521,1,Z0T1LnjxLenygIl,windows_chrome,14.127.210.*,216,...,90,0,0,80,832b91d59d0cb5731431653204a76c0e,300,80,bebefa5efe83beee17a3d245e7c5085b,1458,"10006,10024,10031,10052,10057,10063,10083,1011..."
5599,0,4,15,ffcd02157f67a5b8f9674c787b7f66a1,20130606154517713,1,Vhk7PuuTDvNPu-b,windows_chrome,114.83.119.*,79,...,280,2,1,0,77819d3e0b3467fe5c7b16d68ad923a1,300,168,bebefa5efe83beee17a3d245e7c5085b,1458,
5600,0,2,22,ffd18d8da34b791144eb0b2cf09aa726,20130611223507358,1,Vh16OvTbO6FTjmR,other_other,222.178.10.*,275,...,600,2,1,0,cb7c76e7784031272e37af8e7e9b062c,300,61,bebefa5efe83beee17a3d245e7c5085b,1458,"10024,10063,10111,13800,13866"


5602, the magic number

## Combine the cleaned duplicates with the original data

In [31]:
# remove all original rows whose bidid has been tidied 
df_clean = df[~df['bidid'].isin(df_combined['bidid'])].copy()

# append the tidied duplicates back in 
df_clean = pd.concat(
    [df_clean, df_combined],
    ignore_index=True
)

In [35]:
df_clean[df_clean.duplicated()]

,click,weekday,hour,bidid,timestamp,logtype,ipinyouid,useragent,IP,region,...,slotheight,slotvisibility,slotformat,slotprice,creative,bidprice,payprice,keypage,advertiser,usertag


0 rows :)